nav2_util/include/nav2_util/robot_utils.hpp

```diff
@@ -17,60 +18,24 @@
 #define NAV2_UTIL__ROBOT_UTILS_HPP_
 
 #include <string>
-#include <shared_mutex>
+#include <memory>
+#include <algorithm>
 #include <mutex>
-#include <rclcpp/rclcpp.hpp>
-#include <geometry_msgs/msg/pose_with_covariance_stamped.hpp>
-#include <geometry_msgs/msg/twist.hpp>
-#include <nav_msgs/msg/odometry.hpp>
+#include <thread>
 
-namespace nav2_util
-{
-
-class VelocityPublisher
-{
-public:
-  VelocityPublisher(
-    rclcpp::Node::SharedPtr & node,
-    const std::string topic = std::string("/cmd_vel"));
-
-  void publishCommand(const geometry_msgs::msg::Twist & cmd_vel);
-
-private:
-  rclcpp::Publisher<geometry_msgs::msg::Twist>::SharedPtr vel_pub_;
-};
+#include <geometry_msgs/msg/pose_stamped.hpp>
+#include "tf2_ros/buffer.h"
+#include <tf2_geometry_msgs/tf2_geometry_msgs.h>
+#include "rclcpp/rclcpp.hpp"
 
-class RobotStateHelper
+namespace nav2_util
 {
-public:
-  RobotStateHelper(
-    rclcpp::Node::SharedPtr & node,
-    const std::string odom_topic = std::string("/odom"));
-
-  bool getOdometry(nav_msgs::msg::Odometry::SharedPtr & robot_odom);
-
-  bool getCurrentPose(
-    geometry_msgs::msg::PoseWithCovarianceStamped::SharedPtr & robot_pose,
-    const bool use_topic = true);
-
-private:
-  void onPoseReceived(geometry_msgs::msg::PoseWithCovarianceStamped::SharedPtr robot_pose);
-  void onOdomReceived(nav_msgs::msg::Odometry::SharedPtr msg);
-
-  bool getGlobalLocalizerPose(
-    geometry_msgs::msg::PoseWithCovarianceStamped::SharedPtr & robot_pose);
-  bool getTfPose(geometry_msgs::msg::PoseWithCovarianceStamped::SharedPtr & robot_pose);
 
-private:
-  rclcpp::Node::SharedPtr node_;
-  rclcpp::Subscription<geometry_msgs::msg::PoseWithCovarianceStamped>::SharedPtr pose_sub_;
-  rclcpp::Subscription<nav_msgs::msg::Odometry>::SharedPtr odom_sub_;
-  std::shared_timed_mutex state_mutex_;
-  bool initial_pose_received_, initial_odom_received_;
-  geometry_msgs::msg::PoseWithCovarianceStamped::SharedPtr current_pose_;
-  nav_msgs::msg::Odometry::SharedPtr current_odom_;
-};
+bool getCurrentPose(
```

Comment:
should we consider a `RobotPose` object and hide all the tf stuff in it?

Reply:
Sorry, I'm not sure I follow the comment. This function is branched out into a utility and used everywhere position is consumed in the code to hide TF from the application level code. Is there another way you're suggesting hiding it?

I think maybe what you're thinking is that its annoying that I ask for a TF buffer as an input to this function and a class could hide that? Yes, but then each thread in a process would create its own buffer and that's really not good

Reply:
I'm thinking of the latter. We could create a single RobotPoseGetter and pass it along the same way you are doing with the tf buffer.

I know this might seem silly since we're used to seeing ROS objects all over our stack but won't it be great to reach a level of abstraction in which we're writing code using higher-level robotic terms without considering too much their implementation. Just something to think about, and not blocking this PR.

Reply:
> but won't it be great to reach a level of abstraction in which we're writing code using higher-level robotic terms

Actually I prefer it through the weeds. I'm always suspect of other people's wrappers of "ROS core" tools that most people know well enough to treat them like higher-level operations. I like to read a piece of code and for the most part not need to look at 20 files to get a reasonably full picture. the `nav2_utils` package I generally find messy and as you have probably noticed about recent PRs that I'm slow dismantling it as much as reasonably possible